File: Dynamic Programming/Java/Number of Ways to Form a Target String Given a Dictionary.java
// Problem link: https://leetcode.com/problems/number-of-ways-to-form-a-target-string-given-a-dictionary/description
// Time: O(n * w + m * w)
// Space: O(m * w)
class Solution {
    public int numWays(String[] words, String target) {
        int targetSize = target.length(), wordSize = words[0].length(), MOD = 1000000007;
        int[][] countChar = new int[wordSize][26];
        int[][] DP = new int[targetSize + 1][wordSize + 1];

        for(String word : words) {
            for(int index = 0; index < wordSize; index++) {
                countChar[index][word.charAt(index) - 'a']++;
            }
        }

        for(int index = 0; index < wordSize + 1; index++) {
            DP[targetSize][index] = 1;
        }

        for(int tIndex = targetSize - 1; tIndex >= 0; tIndex--) {
            for(int wIndex = wordSize - 1; wIndex >= 0; wIndex--) {
                DP[tIndex][wIndex] = DP[tIndex][wIndex + 1];
                DP[tIndex][wIndex] %= MOD;
                DP[tIndex][wIndex] += (countChar[wIndex][target.charAt(tIndex) - 'a'] * (long)DP[tIndex + 1][wIndex + 1]) % MOD;
                DP[tIndex][wIndex] %= MOD;
            }
        }

        return DP[0][0];
    }
}

File: Dynamic Programming/Python/Number of Ways to Form a Target String Given a Dictionary.py
# Problem link: https://leetcode.com/problems/number-of-ways-to-form-a-target-string-given-a-dictionary/description
# Time: O(n * w + m * w)
# Space: O(m * w)
class Solution:
    def numWays(self, words: List[str], target: str) -> int:
        targetSize, wordSize = len(target), len(words[0])
        countChar = [[0] * 26 for _ in range(wordSize)]
        DP = [[0] * (wordSize + 1) for _ in range(targetSize + 1)]

        for word in words:
            for index, char in enumerate(word):
                countChar[index][ord(char) - ord('a')] += 1
        
        for k in range(wordSize + 1):
            DP[targetSize][k] = 1

        for tIndex in range(targetSize - 1, -1, -1):
            for wIndex in range(wordSize - 1, -1, -1):
                DP[tIndex][wIndex] = DP[tIndex][wIndex + 1]
                char = target[tIndex]
                DP[tIndex][wIndex] += countChar[wIndex][ord(char) - ord('a')] * DP[tIndex + 1][wIndex + 1]
                DP[tIndex][wIndex] %= 10 ** 9 + 7
        
        return DP[0][0]
